File: path.m
%% Author: Leonardo Pasquato
% Brief: compare script which plots both the ttrajectories with enabled and disabled TRC
% It's possible to check:
%   - 2D trajectory
%   - 3D trajectory
%   - Feedrate

%% Loading csv log

log_trc = 'logtrc.csv';
log_notrc = 'lognotrc.csv';

% Load files
opts = detectImportOptions(log_trc);
opts = setvartype(opts, {'n', 'type'}, 'string');
trc = readtable(log_trc, opts);
notrc = readtable(log_notrc, opts);

trc = sortrows(trc, 't_tot');
notrc = sortrows(notrc, 't_tot');

% Extract x, Y, Z, feedrate, total time
x_trc = trc.X;
y_trc = trc.Y;
z_trc = trc.Z;
x_notrc = notrc.X;
y_notrc = notrc.Y;
z_notrc = notrc.Z;
f_trc = trc.feedrate;
f_notrc = notrc.feedrate;
t_trc = trc.t_tot;
t_notrc = notrc.t_tot;

%% 2D plot
figure;
hold on; grid on; axis equal;
xlabel('X [mm]');
ylabel('Y [mm]');
title('CNCPPevo trajectory comparison (2D) between TRC / no TRC');
plot(x_notrc, y_notrc, 'r--', 'LineWidth', 1.5);
plot(x_trc,   y_trc,   'b-',  'LineWidth', 1.5);
legend('no TRC', 'TRC', 'Location', 'best');

%% 3D plot
figure;
hold on; grid on; axis equal;
xlabel('X [mm]');
ylabel('Y [mm]');
zlabel('Z [mm]');
title('CNCPPevo trajectory comparison (3D) between TRC / no TRC');
plot3(x_notrc, y_notrc, z_notrc, 'r--', 'LineWidth', 1.5);
plot3(x_trc,   y_trc,   z_trc,   'b-',  'LineWidth', 1.5);
legend('no TRC', 'TRC', 'Location', 'best');
view(3);

%% Feedrate plot
figure;
hold on; grid on;
xlabel('Total time [s]');
ylabel('Feedrate [mm/min]');
title('CNCPPevo feedrate comparison between TRC / no TRC');
plot(t_notrc, f_notrc, 'r--', 'LineWidth', 1.5);
plot(t_trc,   f_trc,   'b-',  'LineWidth', 1.5);
legend('no TRC', 'TRC', 'Location', 'best');

File: path.R
# Author: Leonardo Pasquato
# Brief: compare script which plots both trajectories with enabled and disabled TRC
# Features:
#   - 2D trajectory
#   - 3D trajectory
#   - Feedrate

library(readr)
library(dplyr)
library(ggplot2)
library(plotly)

log_trc <- '~/Desktop/Workspace/cncppevo/logtrc.csv'
log_notrc <- '~/Desktop/Workspace/cncppevo/lognotrc.csv'

trc <- read_csv(log_trc, col_types = cols(n = col_character(), type = col_character()))
notrc <- read_csv(log_notrc, col_types = cols(n = col_character(), type = col_character()))

trc <- trc %>% arrange(t_tot)
notrc <- notrc %>% arrange(t_tot)

p2d <- ggplot() +
  geom_path(data = notrc, aes(x = X, y = Y), color = "red", linetype = "dashed", size = 1) +
  geom_path(data = trc, aes(x = X, y = Y), color = "blue", size = 1) +
  labs(
    title = "CNCPPevo trajectory comparison (2D) between TRC / no TRC",
    x = "X [mm]",
    y = "Y [mm]"
  ) +
  coord_equal() +
  theme_minimal()

print(p2d)

p3d <- plot_ly() %>%
  add_trace(data = notrc, x = ~X, y = ~Y, z = ~Z,
            type = 'scatter3d', mode = 'lines',
            line = list(color = 'red', dash = 'dash'),
            name = 'no TRC') %>%
  add_trace(data = trc, x = ~X, y = ~Y, z = ~Z,
            type = 'scatter3d', mode = 'lines',
            line = list(color = 'blue'),
            name = 'TRC') %>%
  layout(
    title = "CNCPPevo trajectory comparison (3D) between TRC / no TRC",
    scene = list(
      xaxis = list(title = "X [mm]"),
      yaxis = list(title = "Y [mm]"),
      zaxis = list(title = "Z [mm]")
    )
  )

p3d

p_feed <- ggplot() +
  geom_line(data = notrc, aes(x = t_tot, y = feedrate), color = "red", linetype = "dashed", size = 1) +
  geom_line(data = trc, aes(x = t_tot, y = feedrate), color = "blue", linewidth = 1) +
  labs(
    title = "CNCPPevo feedrate comparison between TRC / no TRC",
    x = "Total time [s]",
    y = "Feedrate [mm/min]"
  ) +
  theme_minimal()

print(p_feed)

